### DISCUSSION TITLE
Need some clarification on behavior's of Sprite

### DISCUSSION THREAD

**Discussion by meilleur102** [Bodmer/TFT_eSPI#2546] [Q&A - Sprites]
Sorry in advance, it will be a long one. 

I've been playing with the Sprite a lot and I need some clarification on some thing that I can't do.

1) Lagging when the Sprite is not called from the main.cpp.

Function in the Main
```cpp
TFT_eSprite pitSprite = TFT_eSprite(&tft);

void GetPitLimiter(bool isOn);

void setup()
{
  Serial.begin(115200);
  tft.begin();
  tft.setRotation(1);
  tft.fillScreen(TFT_BLACK);
  pitSprite.createSprite(164, 209);
}

void loop()
{
  delay(500)
  GetPitLimiter(true);
  delay(500);
  GetPitLimiter(false);
}


void GetPitLimiter(bool isOn)
{
  int x = 19;
  int y = 113;
  int w = 164;
  int h = 209;
  if (isOn)
  {
    pitSprite.setFreeFont(C24B);
    int tw = tft.textWidth("PIT LIM");

    pitSprite.fillSprite(TFT_YELLOW);
    pitSprite.setTextColor(TFT_BLACK);
    pitSprite.setTextDatum(MC_DATUM);
    pitSprite.drawString("PIT LIM", w / 2, 20);
    pitSprite.pushSprite(x, y);
  }
  if (!isOn && pitSprite.created())
  {
    pitSprite.fillSprite(TFT_BLACK);
    pitSprite.pushSprite(x, y);
  }
}

```

Function in another cpp File

Main

```cpp
TFT_eSprite pitSprite = TFT_eSprite(&tft);

void setup()
{
  Serial.begin(115200);
  tft.begin();
  tft.setRotation(1);
  tft.fillScreen(TFT_BLACK);
  pitSprite.createSprite(164, 209);
}

void loop()
{
  delay(500)
  GetPitLimiter(pitSprite, true);
  delay(500);
  GetPitLimiter(pitSprite,false);
}

```

iRacing.cpp

```cpp
void GetPitLimiter(TFT_eSprite pitSprite, bool isOn)
{
  int x = 19;
  int y = 113;
  int w = 164;
  int h = 209;
  if (isOn)
  {
    pitSprite.setFreeFont(C24B);
    int tw = tft.textWidth("PIT LIM");

    pitSprite.fillSprite(TFT_YELLOW);
    pitSprite.setTextColor(TFT_BLACK);
    pitSprite.setTextDatum(MC_DATUM);
    pitSprite.drawString("PIT LIM", w / 2, 20);
    pitSprite.pushSprite(x, y);
  }
  if (!isOn && pitSprite.created())
  {
    pitSprite.fillSprite(TFT_BLACK);
    pitSprite.pushSprite(x, y);
  }
}


```
When everything is inside the Main loop everything work fine. But when I push the sprite from another class, I have some lag and even some reset from the board. 

2) Is there a way to delete a sprite if it was not created in the same function? I try to delete the sprite to free some memory but when it's not created in the same function i'm not able to delete it succesfully. 
	I try with the pitSprite.created() function, but still not working for me. 
3) Ive search in the Example but I was not able to find it. Can you resize a sprite.
4) Ive experiment with the setColorDepth(1). As I saw in the example, i'ts writen that you can use any 2 color of your choice. I had a sprite with the text black and the fill yellow. But the fill was white. IDK what I miss here. 

Thanks.

**Comment by mcbp223** [comment]
Hi, I think I've been bitten by this too, can you try changing the argument in iRacing.cpp from object copy to reference by adding an ampersand between TFT_eSprite and pitSprite, it should be `void GetPitLimiter(TFT_eSprite& pitSprite, bool isOn)`.

In my case I was trying to use sprites inside STL containers, but the copy constructor kicks in when shuffling them around, and the destructor will delete allocated memory twice, since cloning is not implemented. I've tested them with these additions, @Bodmer could you include this?

in Sprite.h
```cpp
#if __cplusplus >= 201103L
  TFT_eSprite(const TFT_eSprite&) = delete;   // disable cloning, not implemented
  TFT_eSprite(TFT_eSprite&& source) noexcept; //  enable moving instead
#endif
```
and in Sprite.cpp
```cpp
#if __cplusplus >= 201103L
TFT_eSprite::TFT_eSprite(TFT_eSprite&& source) noexcept {
  // move resource ownership from source to *this

  // 1. transfer everything from source
  memcpy(this, &source, sizeof(*this)); // TFT_eSprite must be trivially copyable

  // 2. erase source by calling constructor on its own address
  new (&source) TFT_eSprite(_tft);
  // alternative: memset(&source, 0, sizeof(source));
}
#endif
```

**Reply by meilleur102** [reply]
Good thanks, ill try this.
